### DISCUSSION TITLE
how do the tests work ?

### DISCUSSION THREAD

**Discussion by yamawe** [TheAlgorithms/Python#8791] [Q&A]
Hello, I created this discussion because I don't really understand how automatic tests work. I cloned the directory, launched a virtual machine with python3.11.4, downloaded requirements.txt, pre-commit, black, ruff and mypy. 

After that, I ran "ruff ." which didn't report any errors, when I ran "pre-commit run --all-files --show-diff-on-failure" all the tests passed, but when I ran "mypy .", I got hundreds of errors even though I didn't change anything in the programs. I don't understand how this is possible, since before each merge the pull request should have passed the tests. What's more, when I look at the pull requests, I see that they have passed the tests, whereas the hundreds of programs on which I see mypy errors have not been modified. 

I guess the errors you get when you do 'mypy ...', just ignore them. So can you tell me what tests I need to run on my PC before making a pull request so that I don't get any failed tests?

thank you for your attention

**Comment by CaedenPH** [comment]
@yamawe When `mypy` is ran in isolation it seems to give different results than when ran with `pre-commit` (See #8070)
In order to make a successful pull request, you just need 1)`pre-commit` to pass and 2) `pytest .` to pass

**Reply by yamawe** [reply]
Thank you. Following your message, I launched pytest and received an error concerning an API key: "ERROR web_programming/currency_converter.py - KeyError: "API key must be provided in the 'AMDOREN_API_KEY' environment variable." Did I do something wrong or is this a bug?

**Reply by CaedenPH** [reply]
@yamawe That is not your fault, its a bug (#8738) that i've been trying to get @cclauss to review in order to fix

**Reply by yamawe** [reply]
thank you very much.

**Comment by CaedenPH** [comment]
@yamawe Do you have any more questions?

**Reply by yamawe** [reply]
No, thank you for your help.
